app/RouteHandlers/PayPalNotificationHandler.php

```diff
@@ -55,6 +62,25 @@ private function getValueFromCustomVars( string $customVars, string $key ): stri
 		return !empty( $vars[$key] ) ? $vars[$key] : '';
 	}
 
+	private function requestIsForPaymentCompletion( ParameterBag $post ): bool {
+		return $this->isSuccessfulPaymentNotification( $post->get( 'payment_status' ) ) || (
+				$this->isForRecurringPayment( $post->get( 'txn_type' ) ) &&
+				!$this->isRecurringPaymentCompletion( $post->get( 'txn_type' ) )
```

Comment:
I like how this is split, though the specification of the get value here now mixes levels of abstraction. This can be avoided by passing $post to the methods.